controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -159,81 +162,117 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                                                           final String stream,
                                                           final OperationContext context) {
         Timer timer = new Timer();
+        Map<String, Long> writerTimes = new HashMap<>();
+        Map<String, UUID> writerTimesTxId = new HashMap<>();
+        Map<UUID, String> txnIdToWriterId = new HashMap<>();
+        Map<String, Map<Long, Long>> writerPositions = new HashMap<>();
         return streamMetadataStore.getVersionedState(scope, stream, context, executor)
                 .thenComposeAsync(state -> {
                     final AtomicReference<VersionedMetadata<State>> stateRecord = new AtomicReference<>(state);
 
                     CompletableFuture<VersionedMetadata<CommittingTransactionsRecord>> commitFuture =
-                            streamMetadataStore.startCommitTransactions(scope, stream, MAX_TRANSACTION_COMMIT_BATCH_SIZE, 
-                                    context, executor)
-                            .thenComposeAsync(versionedMetadata -> {
-                                if (versionedMetadata.getObject().equals(CommittingTransactionsRecord.EMPTY)) {
-                                    // there are no transactions found to commit.
-                                    // reset state conditionally in case we were left with stale committing state from
-                                    // a previous execution
-                                    // that died just before updating the state back to ACTIVE but after having 
-                                    // completed all the work.
-                                    return CompletableFuture.completedFuture(versionedMetadata);
-                                } else {
-                                    int txnEpoch = versionedMetadata.getObject().getEpoch();
-                                    List<UUID> txnList = versionedMetadata.getObject().getTransactionsToCommit();
+                            streamMetadataStore.startCommitTransactions(scope, stream, MAX_TRANSACTION_COMMIT_BATCH_SIZE, context, executor)
+                                    .thenApply(commitTxnRecord -> {
```

Comment:
You can merge this `thenApply` into the next `thenComposeAsync` - fewer callbacks will improve your app performance.